packages/backend-cli/src/directLine.ts

```diff
@@ -0,0 +1,772 @@
+      if (options.token && options.conversationId)
+        this.streamUrl = options.streamUrl;
+      else
+        console.warn("streamUrl was ignored: you need to provide a token and a conversationid");
+    }
+    if (options.pollingInterval !== undefined)
+      this.pollingInterval = options.pollingInterval;
+
+    this.activity$ = (this.webSocket
+      ? this.webSocketActivity$()
+      : this.pollingGetActivity$()
+    ).share();
+  }
+
+  // Every time we're about to make a Direct Line REST call, we call this first to see check the current connection status.
+  // Either throws an error (indicating an error state) or emits a null, indicating a (presumably) healthy connection
+  private checkConnection(once = false) {
+    let obs =  this.connectionStatus$
+    .flatMap(connectionStatus => {
+      if (connectionStatus === ConnectionStatus.Uninitialized) {
+        this.connectionStatus$.next(ConnectionStatus.Connecting);
+
+        //if token and streamUrl are defined it means reconnect has already been done. Skipping it.
+        if (this.token && this.streamUrl) {
+          this.connectionStatus$.next(ConnectionStatus.Online);
+          return Observable.of(connectionStatus);
+        } else {
+          return this.startConversation().do(conversation => {
+            this.conversationId = conversation.conversationId;
+            this.token = this.secret || conversation.token;
+            this.streamUrl = conversation.streamUrl;
+            this.referenceGrammarId = conversation.referenceGrammarId;
+            if (!this.secret)
+              this.refreshTokenLoop();
+
+            this.connectionStatus$.next(ConnectionStatus.Online);
+          }, error => {
+            this.connectionStatus$.next(ConnectionStatus.FailedToConnect);
+          })
+          .map(_ => connectionStatus);
+        }
+      }
+      else {
+        return Observable.of(connectionStatus);
+      }
+    })
+    .filter(connectionStatus => connectionStatus != ConnectionStatus.Uninitialized && connectionStatus != ConnectionStatus.Connecting)
+    .flatMap(connectionStatus => {
+      switch (connectionStatus) {
+        case ConnectionStatus.Ended:
+          return Observable.throw(errorConversationEnded);
+
+        case ConnectionStatus.FailedToConnect:
+          return Observable.throw(errorFailedToConnect);
+
+        case ConnectionStatus.ExpiredToken:
+          return Observable.throw(errorExpiredToken);
+
+        default:
+          return Observable.of(null);
+      }
+    })
+
+    return once ? obs.take(1) : obs;
+  }
+
+  private expiredToken() {
+    const connectionStatus = this.connectionStatus$.getValue();
+    if (connectionStatus != ConnectionStatus.Ended && connectionStatus != ConnectionStatus.FailedToConnect)
+      this.connectionStatus$.next(ConnectionStatus.ExpiredToken);
+  }
+
+  private startConversation() {
+    //if conversationid is set here, it means we need to call the reconnect api, else it is a new conversation
```

Comment:
// if conversationId is set we need to reconnect to an existing conversation otherwise start a new one. Should these be different methods so it is explicit what we want instead of being implicit based on the current state of the object?